### PR TITLE
Insertsmcalls

### PR DESCRIPTION
**Pull request by vext01**
This adds the stackmap insertion pass that Lukas need for his stack reconstruction work.

See commit messages for more info.

A companion PR will be needed for ykjit/yk -- coming soon.

### CODE REVIEW

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    // instructions of all successor blocks.
+    for (unsigned SuccIdx = 0; SuccIdx < Term->getNumSuccessors(); SuccIdx++)
+      SuccInsts.insert(&*Term->getSuccessor(SuccIdx)->begin());
+  }
+  return SuccInsts;
+}
+
+/// Replaces the value set behind the pointer `S` with the value set `R` and
+/// returns whether the set behind `S` changed.
+bool LivenessAnalysis::updateValueSet(std::set<Value *> *S,
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:63]: This method name (indeed the entire method!) feel weird. Can we find something more descriptive?

Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:63]: Maybe, but why is it weird? It updates a `set` of `Value *`, hence `updateValueSet()`.

Is the confusion in `update`? How about `replaceValueSet()`?

(Note that this is not new code)

Comment (ltratt) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:63]: It's the "Replaces the value set behind the pointer `S`" bit -- it feels rather non-idiomatic. Maybe there's no other way, but it just feels... weird.

Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:63]: Let me see if I can remove the need for this oddity.

Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: Does [a3a39a5](https://github.com/ykjit/ykllvm/pull/38/commits/a3a39a573127caf986087e883d5f77b27e6d6d84) work better?

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 2]
```diff
@@ -138,22 +136,20 @@ LivenessAnalysis::LivenessAnalysis(Function *Func) {
         for (Instruction *SI : SuccInsts) {
           NewOut.insert(In[SI].begin(), In[SI].end());
         }
-        Changed |= updateValueSet(&Out[I], std::move(NewOut));
```
Comment (vext01) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: wait, I didn't notice the `move`. They can go now.

Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: Fixed in eb0c89b66c3b. Retested to be sure it didn't break too.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@ void vset_union(const std::set<Value *> &S1, const std::set<Value *> &S2,
 }
 
 namespace llvm {
-/// Find the successor instructions of the specified instruction.
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: This shouldn't have been deleted shouldit?

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp [hunk 4]
```diff
@@ -58,12 +58,10 @@ LivenessAnalysis::getSuccessorInstructions(Instruction *I) {
   return SuccInsts;
 }
 
-/// Replaces the value set behind the pointer `S` with the value set `R` and
-/// returns whether the set behind `S` changed.
-bool LivenessAnalysis::updateValueSet(std::set<Value *> *S,
-                                      const std::set<Value *> R) {
-  const bool Changed = (*S != R);
-  *S = R;
+bool LivenessAnalysis::updateValueSet(std::set<Value *> &S,
```
Comment (ltratt) [inline llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:61]: I think this function still needs *a* comment.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 1 above)
Comment (ltratt) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: Apart from the disappearing doc comments, this does feel a bit nicer.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 3 above)
Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: In the commit message i talked about removing the duplication of the docstrings based on an answer to  stackoverflow question (found by searching for "should i document c++ in the header or the cpp file :P).

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 4 above)
Comment (vext01) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp:61]: It's in the header.

File: llvm/lib/Transforms/Yk/LivenessAnalysis.cpp (hunk 3 above)
Comment (ltratt) [reply llvm/lib/Transforms/Yk/LivenessAnalysis.cpp]: Ah I see what that means now!

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

**Pull request merged**